Make optimization based on the entire posterior and not on the marginal mean parameters.

cetagostini opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

## Description
<!--- Describe your changes in detail -->
Structural change on how responses are computed in the optimizer, we are now using the entire "posterior" of the model, and additionally we are creating a new notebook where we can encode certain risk levels.

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
<!--- Please list the modules that are affected by this PR by typing an `x` in the boxes below: -->
- [x] MMM
- [ ] CLV
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--1151.org.readthedocs.build/en/1151/

<!-- readthedocs-preview pymc-marketing end -->

wd60622 commented on pymc_marketing/mmm/mmm.py:
```diff
@@ -1613,6 +1614,49 @@ def format_recovered_transformation_parameters(
+            'channels': ['x1', 'x2']
+        }
+        """
+        if self.idata is not None:
+            saturation_params: dict[str, np.ndarray] = {
+                key: self.idata.posterior[f"saturation_{key}"].values
+                for key in self.saturation.default_priors.keys()
+            }
+            adstock_params: dict[str, np.ndarray] = {
+                key: self.idata.posterior[f"adstock_{key}"].values
+                for key in self.adstock.default_priors.keys()
+            }
+        else:
+            raise ValueError("idata is not initialized.")
```
Use `self.fit_results` here instead. It already catches and raises if the model is not fit

wd60622 commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -98,32 +112,61 @@ def objective(self, budgets: list[float]) -> float:
+        sig = inspect.signature(self.objective_function)
+        params = sig.parameters
+
+        # Prepare arguments
+        function_kwargs = self.objective_function_kwargs.copy()
+        if "assets" in params:
+            function_kwargs["assets"] = budgets
+
+        if "capital" in params:
+            function_kwargs["capital"] = np.sum(budgets)
```
This isn't scalable. What happened with the Callable[[np.ndarray], float]? Maybe we have `Callable[[np.ndarray, np.ndarray], float]` where the second parameter are the budgets

wd60622 commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -217,8 +262,420 @@ def allocate_budget(
             raise MinimizeException(f"Optimization failed: {result.message}")
+
+
+class RiskAssessment:
```
Thought we were thinking about the functional approach over the staticmethods? 

```python
# Option 1
def create_tail_distance(confidence_level: float):
    def tail_distance(samples) -> float: 
        ...
    return tail_distance

objective = create_tail_distance(confidence_level=...)

# Option 2
def tail_distance(samples, confidence_level: float) -> float:
     ...

from functools import partial
objective = partial(tail_distance, confidence_level=...)
```

wd60622 commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -56,6 +57,10 @@ class BudgetOptimizer(BaseModel):
         Default is True.
+    objective_function : Callable[[np.ndarray], float], optional
+        The objective function to maximize. Default is the mean of the response distribution.
+    objective_function_kwargs : dict, optional
+        Additional keyword arguments for the objective function. Default is an empty dictionary.
```
Delete this 

```suggestion
```

wd60622 commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -98,32 +109,48 @@ def objective(self, budgets: list[float]) -> float:
+
+        # Dynamically reshape parameters in `second_params` to match `_response`
+        for param_name, param_value in second_params.items():
+            if param_value.ndim == 3:  # Check if it lacks the `time` dimension
+                # Reshape by adding a singleton dimension for broadcasting over time
+                second_params[param_name] = param_value.reshape(
+                    param_value.shape[0], param_value.shape[1], 1, param_value.shape[2]
+                )
```
Can we make an issue for this if our plan is to extend.

cetagostini replied:
Sure, let me open an issue!

wd60622 commented on pymc_marketing/mmm/mmm.py:
```diff
@@ -2166,10 +2207,10 @@ def allocate_budget_to_maximize_response(
         noise_level : float, optional
             The level of noise added to the allocation strategy (by default 1%).
+        objective_function : ObjectiveFunction, optional
+            The objective function to maximize.
```
Mention the default behavior here

cetagostini replied:
Done!

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""Risk assessment module."""
```
Can you add a bit more to this and expand on what the samples and budgets are in the signature. Maybe a bit more in the samples specifically, i.e. shape, meaning

cetagostini replied:
- Will change because name will change.

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+ObjectiveFunction = Callable[[np.ndarray, np.ndarray], float]
+
+
+def default_assessment(samples: np.ndarray, budgets: np.ndarray):
```
Think `average_response` might be a better name. Then in the documentation, we can say that this is the default assessment of the response curve.

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    The Diversification Ratio is calculated as:
+
+    .. math::
+        DR = \frac{\\sum_{i=1}^{n} w_i \\cdot \\sigma_i}{\\sigma_p}
```
This is not rendering well in the documentation. I think it has to dow ith the double backslash and R string. \\sum -> \sum, etc 

Similar to #1085

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    ObjectiveFunction
+        A function that calculates the tail distance metric given samples and budgets.
+    """
+
```
add confidence level check here like in the other function

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    """
+
+    def _value_at_risk(samples: np.ndarray, budgets: np.ndarray) -> float:
+        if not 0 < confidence_level < 1:
+            raise ValueError("Confidence level must be between 0 and 1.")
```
This check can be done before creating the inner function

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    """
+
+    def _conditional_value_at_risk(samples: np.ndarray, budgets: np.ndarray) -> float:
+        if not 0 < confidence_level < 1:
```
same here

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    ------
+    ValueError
+        If confidence_level is not between 0 and 1.
+        If no samples fall below the VaR threshold.
```
This technically isn't raised by this function but by the inner function so I would remove

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    ObjectiveFunction
+        A function that calculates the Sharpe Ratio given samples and budgets.
+
+    Raises
+    ------
+    ValueError
+        If the standard deviation of excess returns is zero.
```
Same thing here. remove

cetagostini replied:
Removed.

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    ObjectiveFunction
+        A function that calculates the RAROC value given samples and budgets.
+
+    Raises
+    ------
+    ValueError
+        If `capital` is less than or equal to zero, as RAROC requires positive capital.
```
Same note here

cetagostini replied:
Removed.

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+    def _adjusted_value_at_risk_score(
+        samples: np.ndarray, budgets: np.ndarray
+    ) -> float:
+        if not 0 <= risk_aversion <= 1:
+            raise ValueError("Risk aversion parameter must be between 0 and 1.")
+        if not 0 < confidence_level < 1:
+            raise ValueError("Confidence level must be between 0 and 1.")
```
These can be moved before creating the inner function

wd60622 commented on pymc_marketing/mmm/risk_assessment.py:
```diff
@@ -0,0 +1,461 @@
+        excess_returns = samples - risk_free_rate
+        mean_excess_return = np.mean(excess_returns)
+        std_excess_return = np.std(excess_returns, ddof=1)
+        if std_excess_return == 0:
+            raise ValueError(
+                "Standard deviation of excess returns is zero; Sharpe Ratio is undefined."
+            )
```
This is have an issue with floating point precision. Let's remove or check differently

wd60622 commented on tests/mmm/test_risk_assessment.py:
Let's not use random values for these tests but rather define arrays

cetagostini replied:
Create a test checking expected output for the array.

juanitorduz commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -98,32 +109,48 @@ def objective(self, budgets: list[float]) -> float:
+        response = second_transform.function(x=_response, **second_params).eval()
+        return response.sum(axis=(2, 3)).flatten()  # dim: (chain * draw)
+
+    def objective(self, budgets):
```
missing type hints

juanitorduz replied:
and docstrings

juanitorduz commented on pymc_marketing/mmm/mmm.py:
```diff
@@ -1613,6 +1614,46 @@ def format_recovered_transformation_parameters(
         return channels_info
 
+    def _format_parameters_for_budget_allocator(
+        self,
```
If re remove the `,` I think this function signature first in one line ;)

juanitorduz commented:
This is super cool @cetagostini ! I really learned a lot from the new notebook 🚀 .

I left some comments :)

juanitorduz commented:
(I will give a more detailed review in the up-coming days)

wd60622 commented:
In the new notebook, can we do `from pymc_marketing.mmm import risk_assessment` instead. That is a bit more convention.

cluhmann commented:
I have not had a chance to get through the PR yet, but can I ask how settled we are on `risk_assessment` as a name?  Are we really **assessing** risk? To me, all of this stuff is just constructing an utility/objective/loss function.  Just wanted to throw this out there before it all got merged. I will also try to review the PR more thoroughly.

cetagostini commented:
> I have not had a chance to get through the PR yet, but can I ask how settled we are on `risk_assessment` as a name? Are we really **assessing** risk? To me, all of this stuff is just constructing an utility/objective/loss function. Just wanted to throw this out there before it all got merged. I will also try to review the PR more thoroughly.

I follow your point, and agree with part of it. I think the core of what happens is that by changing different loss functions, and doing so, we could or could not make an assessment on how risky a certain budget is, of course depends on how the loss function is design to assess a point estimate or do something more.

That said, the risk-oriented name still holds merit because a substantial part of the functionality (e.g., Sharpe ratio, diversification ratio, VaR) is directly tied to risk evaluation and management, even if these objective functions feed into broader optimization tasks. The module is create with only function that assess not only returns but the uncertainty of those returns.

Additionally, since it is linked to marketing, I think part of doing allocations in marketing is considering their potential risk. That is part of marketing slang, in the same way that we say response curve and not diminishing returns curves, or we say adstock and not lagging effects.

Let me know your thoughts, and I’m happy to adjust the naming if it better reflects the module!

cetagostini commented:
Hey I took your feedback into account and changed the optimizer to be based on numpyro. That means that the functions in risk assessment have to be written in numpyro.

I had to redo some functions that I didn't find in numpyro, but I also added tests to see that they behave the same as in numpy.

cetagostini commented:
_Quick mention_: I found some issues in the sample posterior prediction that I believe affect the notebook results, making all the distributions look very similar and perhaps blurring the point. I'll do some testing and see if I can figure it out.

cetagostini commented:
All notebooks update to use the new optimization notebook.

cetagostini commented:
Ready for final review, everything should be done.

Cc: @juanitorduz @cluhmann

juanitorduz reviewed (changes requested):
@cetagostini can we revert the examples in the example notebook? This seems like an old version and has many errors and typos 🙏

juanitorduz commented on pymc_marketing/mmm/budget_optimizer.py:
```diff
@@ -182,43 +273,44 @@ def allocate_budget(
-                "method": "SLSQP",
-                "options": {"ftol": 1e-9, "maxiter": 1_000},
-            }
+            minimize_kwargs = self.DEFAULT_MINIMIZE_KWARGS
```
Can we consider https://github.com/pymc-labs/pymc-marketing/issues/1193 here?

cetagostini replied:
Like it, will apply it.

cetagostini replied:
Mypy doesn't like:
```python
        if minimize_kwargs is None:
            minimize_kwargs = {
                **self.DEFAULT_MINIMIZE_KWARGS,
                **(minimize_kwargs or {}),
            }
```

Change to 
```
if minimize_kwargs is None:
            minimize_kwargs = self.DEFAULT_MINIMIZE_KWARGS.copy()
        else:
            minimize_kwargs = {**self.DEFAULT_MINIMIZE_KWARGS, **minimize_kwargs}
```

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Utility functions for Bayesian optimization.
```
shall we call this file `utils.py` for consistency across the package?

juanitorduz replied:
Actually there are many functions here which we can move to their own module `risk.py`. This way we keep the modules more consistent instead of putting many thing here. wdyt?

cetagostini replied:
Yes, talking with @cluhmann and @wd60622 we think that even though most functions are now directed at risk, the truth is that the change is bigger than that. We are allowing utility functions to be changed for any purpose, one could be risk, but the user could define anything. Therefore, calling it **risk** or something is really like closing the scope when it could be much more.

I think _utils_ would not be a suitable name, since we are talking about **utility functions**, which is a bayesian name, basically a similar term for the objective function in the bayesian optimization context. Those functions are not simple `utils` for everything.

What do you think?

wd60622 replied:
I am for `utility.py`. I read that different than `utils.py`

juanitorduz replied:
Ok! Ni strong opinion here. In a different PR we just need to harmonize the term across the package if needed :)

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+
+import pytensor.tensor as pt
+
+UtilityFunction = Callable[[pt.TensorVariable, pt.TensorVariable], float]
```
Shall we call this `UtilityFunctionType`? Otherwise it could be interpreted as a class

cetagostini replied:
Yes, we can do that!

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+    return (1 - weight) * sorted_x[idx_floor] + weight * sorted_x[idx_ceil]
+
+
+def average_response(
+    samples: pt.TensorVariable, budgets: pt.TensorVariable
+) -> pt.TensorVariable:
+    """Compute the average response of the posterior predictive distribution."""
+    return pt.mean(samples)
```
Do we really need this function? Note we are not using the `budget` variable at all, right?

cetagostini replied:
We need the same signature in all functions in order to make compatible an scalable. I had an implementation that was more directed per function but was taking a few flexibility from users, with @wd60622 we arrive to this conclusion with makes everything more flexible, the only inconvenience is we need `samples` and `budgets` across all, even when we don't use them.

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+        raise ValueError("Confidence level must be between 0 and 1.")
+
+    def _tail_distance(
+        samples: pt.TensorVariable, budgets: pt.TensorVariable
```
we are not using the `budgets` variable, so we should remove it right?

cetagostini replied:
Replied before!

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+    ----------
+    alpha : float, optional
+        Risk tolerance parameter (default is 0.5).
+    confidence_level : float, optional
```
We should add "confidence level must be between 0 and 1."

cetagostini replied:
Yes, good catch!

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+
+    Parameters
+    ----------
+    confidence_level : float, optional
```
We should add "confidence level must be between 0 and 1."

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+
+    Parameters
+    ----------
+    confidence_level : float, optional
```
We should add "confidence level must be between 0 and 1."

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+
+    Parameters
+    ----------
+    confidence_level : float, optional
```
We should add "confidence level must be between 0 and 1."

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+    ----------
+    .. [1] Rockafellar, R.T., & Uryasev, S. (2000). Optimization of Conditional Value-at-Risk.
+    """
+    if not 0 < confidence_level < 1:
```
any problems if is exactly zero or one?

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+
+    Parameters
+    ----------
+    confidence_level : float, optional
```
We should add "confidence level must be between 0 and 1."

juanitorduz commented on tests/mmm/test_budget_optimizer.py:
```diff
@@ -28,38 +28,50 @@
-                "channel_1": {
-                    "adstock_params": {"alpha": 0.5},
-                    "saturation_params": {"lam": 10, "beta": 0.5},
+                "saturation_params": {
```
should we still test the scalar case?

cetagostini replied:
What do you mean here? What scalar case?

juanitorduz commented on tests/mmm/test_utility.py:
```diff
@@ -0,0 +1,376 @@
+
+rng: np.random.Generator = np.random.default_rng(seed=42)
+
+EXPECTED_RESULTS = {
```
where are these values coming from?

cetagostini replied:
I ran a notebook to check what would be the results for each function given the parameters, this test validates they return those values every time, meaning, function behavior its not changing and return consistent results.

juanitorduz commented on pymc_marketing/mmm/utility.py:
```diff
@@ -0,0 +1,546 @@
+    UtilityFunction
+        A function that calculates the tail distance metric given samples and budgets.
+    """
+    if not 0 < confidence_level < 1:
```
We should add a test fro these `ValueErrors`

juanitorduz replied:
Alternatively, we could use ` @validate_call` from pydantic as in https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/mmm/mmm.py#L74

cetagostini replied:
Agree!

cetagostini replied:
Test in place!

juanitorduz reviewed (changes requested):
I tried to give a detailed look and left some comments 🙏

juanitorduz commented:
**Important Comments:**

- Before we merge, we need to make sure the changes in the notebooks are meaningful. For example, the MMM Example version seems old as it reverts typo corrections and does not use the `Prior` class. We should only change the notebooks relevant to this PR.
Can you please re-run the MMM case study and expand on the budget allocation section a bit to showcase the gains of this new approach and how to use it in practice?

Thank you for this great and ambitious PR, @cetagostini ! These suggestions are to make sure we deliver a great solution 💪 🙇

cetagostini commented:
> **Important Comments:**
> 
> * Before we merge, we need to make sure the changes in the notebooks are meaningful. For example, the MMM Example version seems old as it reverts typo corrections and does not use the `Prior` class. We should only change the notebooks relevant to this PR.
>   Can you please re-run the MMM case study and expand on the budget allocation section a bit to showcase the gains of this new approach and how to use it in practice?
> 
> Thank you for this great and ambitious PR, @cetagostini ! These suggestions are to make sure we deliver a great solution 💪 🙇

Thank you for the detailed review, I need some feedback on a few comments, I'll proceed with the rest 🙌🏻

PS: Sorry for the MMM example, I got issues with the notebook and maybe the rebase revert those, let me bring back the one from main.

juanitorduz commented:
Thanks @cetagostini ! Git and notebooks is a pain!👌🙇

(btw cool pytensor stuff... I am learning quite a lot from these risks metrics!)

cetagostini commented:
@juanitorduz @wd60622 changes applied.

juanitorduz reviewed (approved):
Thanks @carlosagostini ! This is a great addition! I think we should merge this one and iterate. 🚀
